lib/backend/postgres/tx.go

```diff
@@ -0,0 +1,485 @@
+func (tx *pgTx) Err() error {
+	return tx.err
+}
+
+// Commit the transaction. The same error returned from the Err method is
+// returned from Commit when the transaction is in an error state.
+func (tx *pgTx) Commit() error {
+	if tx.err != nil {
+		return tx.err
+	}
+	tx.err = convertError(tx.sqlTx.Commit())
+	return tx.err
+}
+
+// Rollback the transaction with an error. The error passed to Rollback is
+// converted to a trace error and set as the transaction error returned from
+// Err. If the transaction is already in an error state, the error is
+// overridden by the error passed. Passing a nil error is considered a bug,
+// but the rollback will continue with a generated error if the transaction
+// is not already in an error state.
+func (tx *pgTx) Rollback(err error) error {
+	switch {
+	case err != nil && tx.err == nil:
+		// Standard path. Rollback with error.
+		tx.rollback(err)
+
+	case err != nil && tx.err != nil:
+		// Tx in failed state. Override error (e.g. CompareAndSwap).
+		tx.err = err
+
+	case err == nil && tx.err == nil:
+		// Tx OK. Rollback called with nil err. Rollback since that was the intent.
+		tx.rollback(trace.BadParameter("[BUG] Rollback called with nil error."))
+		tx.cfg.Log.Debug(tx.err)
+
+	case err == nil && tx.err != nil:
+		// Tx in failed state. Rollback called with nil error. Override error to signal issue.
+		tx.err = trace.BadParameter("[BUG] Rollback called with nil error on failed transaction.")
+		tx.cfg.Log.Debug(tx.err)
+	}
+	return trace.Wrap(tx.err)
+}
+
+// rollback and return true when err is not nil.
+//
+// This method is a control-flow helper for the other pgTx methods. Accepting
+// nil errors and returning true or false improves readability in the same
+// manner as trace.Wrap. Its name is purposefully short and discreet.
+func (tx *pgTx) rollback(err error) bool {
+	if err == nil {
+		return false
+	}
+	if e := tx.sqlTx.Rollback(); e != nil {
+		tx.cfg.Log.WithError(err).Errorf("Failed to rollback: %v", e)
+	}
+	tx.err = convertError(err)
+	return true
+}
+
+// DeleteEvents created before expiryTime.
+func (tx *pgTx) DeleteEvents(expiryTime time.Time) {
+	if tx.err != nil {
+		return
+	}
+
+	const query = `DELETE FROM event WHERE created < $1`
+	_, err := tx.sqlTx.ExecContext(tx.ctx, query, expiryTime)
+	tx.rollback(err)
+}
+
+// DeleteExpiredLeases removes leases whose expires column is not null and is
+// less than the current time.
+func (tx *pgTx) DeleteExpiredLeases() {
+	if tx.err != nil {
+		return
+	}
+
+	const query = `DELETE FROM lease WHERE (expires IS NOT NULL AND expires < $1)`
+	_, err := tx.sqlTx.ExecContext(tx.ctx, query, tx.now())
+	tx.rollback(err)
+}
+
+// DeleteItems not referencing an event or a valid lease.
+func (tx *pgTx) DeleteItems() {
+	if tx.err != nil {
+		return
+	}
+
+	const query = `
+		DELETE FROM item WHERE (key, id) IN (
+			SELECT key, id
+			FROM item
+			LEFT JOIN lease USING (key, id)
+			LEFT JOIN event USING (key, id)
+			WHERE event.key IS NULL
```

Comment:
Not sure if I missing something but  how `key` value  can be set to nil in `event` table ?